rubert/src/bin/validate.rs

```diff
@@ -0,0 +1,356 @@
+        let talks = config.data.talks;
+        let skip = match config.data.range.start_bound() {
+            Bound::Included(start) => *start,
+            Bound::Excluded(start) => start + 1,
+            Bound::Unbounded => 0,
+        };
+        let take = match config.data.range.end_bound() {
+            Bound::Included(end) => end + 1,
+            Bound::Excluded(end) => *end,
+            Bound::Unbounded => 2467,
+        } - skip;
+        let source = Pipeline::build(&config.tokenizer, &config.source);
+        let target = Pipeline::build(&config.tokenizer, &config.target);
+        let errors = Array1::zeros(11);
+
+        Self {
+            talks,
+            skip,
+            take,
+            source,
+            target,
+            errors,
+        }
+    }
+
+    /// Computes the mean of the difference between source and target.
+    fn mean_absolute_error(source: &Embedding2, target: &Embedding2) -> f32 {
+        (source.deref() - target.deref())
+            .mapv(|v| v.abs())
+            .mean()
+            .unwrap_or_default()
+    }
+
+    /// Computes the mean of the difference between source and target relative to the source.
+    fn mean_relative_error(source: &Embedding2, target: &Embedding2) -> f32 {
+        ((source.deref() - target.deref()) / source.deref())
+            .mapv(|v| v.is_finite().then(|| v.abs()).unwrap_or_default())
+            .mean()
+            .unwrap_or_default()
+    }
+
+    /// Computes the mean of the squared difference between source and target.
+    fn mean_squared_absolute_error(source: &Embedding2, target: &Embedding2) -> f32 {
+        (source.deref() - target.deref())
+            .mapv(|v| v.powi(2))
+            .mean()
+            .unwrap_or_default()
+            .sqrt()
+    }
+
+    /// Computes the mean of the squared difference between source and target relative to the source.
+    fn mean_squared_relative_error(source: &Embedding2, target: &Embedding2) -> f32 {
+        ((source.deref() - target.deref()) / source.deref())
+            .mapv(|v| v.is_finite().then(|| v.powi(2)).unwrap_or_default())
+            .mean()
+            .unwrap_or_default()
+            .sqrt()
+    }
+
+    /// Computes the cosine similarity between source and target.
+    fn cosine_similarity(source: &Embedding2, target: &Embedding2) -> f32 {
+        let norms =
+            source.mapv(|v| v.powi(2)).sum().sqrt() * target.mapv(|v| v.powi(2)).sum().sqrt();
+        (norms.is_finite() && norms > 0.0)
+            .then(|| (source.deref() * target.deref() / norms).sum())
+            .unwrap_or_default()
+    }
+
+    /// Computes various errors between source and target embeddings based on the chosen ted talks.
+    fn validate(&mut self) -> &mut Self {
+        let mut reader = Reader::from_path(self.talks).unwrap();
+        let progress = ProgressBar::new(self.take as u64);
+        let mut errors = Array2::<f32>::zeros((330644, 5));
```

Comment:
why `330644`?

Reply:
this is the total number of sentences in all talks.

i added comments in the code to explain all the magic numbers above.